KF-739

misohu opened the pull request:
Create image with CI

jnsgruk commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+FROM ubuntu:22.04
```
We should be using ubuntu from the ECR registry ideally, it's updated more frequently

jnsgruk commented on Dockerfile:
Is there a reason not to use rockcraft for a brand new image?

misohu replied:
The task was planned with Dockerfile ... I think we can convert to ROCK in the future. I think thats the plan also for other images. What do you think @i-chvets ?

jnsgruk replied:
To be honest, we're better off building competency in the team with rockcraft - that *is* your destination. It makes sense to use dockerfiles where we're just reusing upstream stuff, but in general we should be moving toward rockcraft wherever possible.

jnsgruk commented on server.py:
```diff
@@ -0,0 +1,128 @@
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--port",
+        "-p",
+        type=int,
+        help="Port on which run the sync server",
+        default=None,
+        required=False,
+    )
```
Given where this will be run, you might consider allowing this to be set using an environment variable

jnsgruk replied:
Actually, same applies for all args.

ca-scribner replied:
I believe I've used [this pattern](https://stackoverflow.com/questions/10551117/setting-options-from-environment-variables-when-using-argparse) before to get the best of both worlds.  Defaults to the envvar unless there's an arg passed.  Kinda like...viper(??) in go.  Something like that

jnsgruk replied:
Yeh, that's also a good route

jnsgruk commented on server.py:
```diff
@@ -0,0 +1,128 @@
+PORT = os.getenv("PORT", 80)
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
```
This should probably be adjustable - consider using an env var.

jnsgruk commented on server.py:
```diff
@@ -0,0 +1,128 @@
+
+
+def main(port: int, label: str, folder: str) -> None:
+    """
+    Runs a server for injecting resources into namespaces by given label
+    """
```
Good docstring, but pretty sure it wouldn't pass muster with the docstring checks in ruff or pydocstyle using the google styleguide??

ca-scribner replied:
I think this will get tidied up as soon as there's CI applying this repo's `tox -e lint`

misohu replied:
Added lint

jnsgruk commented on server.py:
```diff
@@ -0,0 +1,128 @@
+    """
+    Runs a server for injecting resources into namespaces by given label
+    """
+    logger.info("Resource dyspatcher service alive")
```
```suggestion
    logger.info("Resource dispatcher service alive")
```

jnsgruk commented on server.py:
```diff
@@ -0,0 +1,128 @@
+    """
+    logger.info("Resource dyspatcher service alive")
+    server = server_factory(port, label, folder)
+    logger.info("Serving forever")
```
This is a bit of useless message - what interface? what port? serving what?

jnsgruk commented on server.py:
```diff
@@ -0,0 +1,128 @@
+
+def server_factory(controller_port: int, label: str, folder: str, url: str = "") -> HTTPServer:
+    """
+    Returns an HTTPServer populated with Handler with customized settings
```
nit: Canonical uses British english
```suggestion
    Returns an HTTPServer populated with Handler with customised settings
```

jnsgruk commented on server.py:
```diff
@@ -0,0 +1,128 @@
+
+    class Controller(BaseHTTPRequestHandler):
+        def sync(self, parent, children):
+            logger.info("Got new request")
```
This is not a useful message - please include some context

jnsgruk commented on server.py:
```diff
@@ -0,0 +1,128 @@
+            desired_secret_count = 0
+            desired_resources = []
+            desired_status = {
+                "resources-ready": len(children["Secret.v1"]) == desired_secret_count
+                and "True"
+                or "False"
```
This is a bit code-golfy - consider breaking out the logic to increase readability

jnsgruk commented on tox.ini:
```diff
@@ -0,0 +1,65 @@
+      --skip {toxinidir}/./.mypy_cache \
+      --skip {toxinidir}/./icon.svg --skip *.json.tmpl
+    # pflake8 wrapper supports config from pyproject.toml
+    pflake8 {[vars]src_path}
+    isort --check-only --diff {[vars]all_path}
+    black --check --diff {[vars]all_path}
```
I've moved recently to using `ruff` + `black` which I've  found much more stable in CI - it's also now the recommended path in the charmcraft template. See [this repo](https://github.com/jnsgruk/parca-operator) for example tox/pyproject file with equivalent rules

i-chvets replied:
@jnsgruk This linting process is the same CI template that is used across all KF charms. There is a conversation in-team and with COS to have a new unified CI (spec is in Braindump stage). We can address this in next sprint along with other dev process improvements.

jnsgruk reviewed (commented):
some comments :)

ca-scribner commented on README.md:
```diff
@@ -1,2 +1,2 @@
-# resource-dispatcher-image
-Image for Kubernetes meta controller server which generates Kubernetes resources which need to be injected to user namespaces. 
+# Resource dispatcher Docker image
+This is an image for resource dispatcher.
```
Can we expand this to describe what the tool does?

ca-scribner replied:
This is a great thing to break into a separate PR as well.  It can be standalone

ca-scribner replied:
I agree this is a relatively small PR functionality wise, but that's the point.  There's 29 files changed here, so it looks and reviews like a big PR even though it could be a few small PRs.

imo (and I bet if she was here, in @DnPlas's opinion) we can do much better at splitting PRs and making these reviews simpler.  Adding a README.md as a separate PR is really easy, so I'd prefer it be done that way.  But that isn't a showstopper here.

Making separate PRs sounds harder, and it is harder *after the fact*.  But if done as the work progresses it is easier on everyone.  The boilerplate (copied tox.ini, stuff like that) could be a PR that's reviewed right away.  The actual resource dispatcher script and its tests can be their own PR.  The Dockerfile/CI that builds the image can be a separate PR. These all have distinct functionality that can be reviewed individually

DnPlas replied:
I agree with @ca-scribner. There are a couple of changes that definitely belong in other PRs. I can understand that splitting after submitting a large number of changes can be hard, and it's hard to be fair.

To me changes in this PR could be grouped in:
* Group 1: boilerplate (tox.ini, requirements*, workflows, .gitignore, etc.)
* Group 2: documentation changes
* Group 3: src and tests, and I'd argue these could even be in separate PRs

As @ca-scribner said, this does not have to be a showstopper, but it would be nice.

jnsgruk replied:
Agree - I don't think this is a huge deal and given that there has been a fair bit of review already, probably just one for next time.

One thing you *could* do is rebase this PR into 3 clear commits (according to Daniela's suggestion above) and rebase rather than squash merge, but again shouldn't be a blocker a don't think.

i-chvets replied:
From our best practices:
> verifiable chunks that can function separately is very helpful

Except documentation and workflows, the proposed PR break down will go against exactly that requirement: Group 1 will fail without Group 3. Group 3 will not have way to run tests, because `tox.ini` updates are in Group 1. And Group 1 will not have source code to test.
It is impossible to verify Group 1 without Group 3. May be a review of just changes is clearer, but those two pieces won't function by themselves.
I don't buy argument about test driven development on this subject, because it is not development, but code commit strategy. During development it is perfectly fine to start with `tox.ini` changes to make test fail and start building up functionality from that. Separating code later for commits is counterproductive.
Putting tox/source/tests together gives exactly what we want: a verifiable chunk that can function.
In this particular case of brand new charm submitting separate PRs make even reviewing unproductive, because I would need to review two PRs separately and merge them both to achieve testable commit. So, at the end I will be reviewing two PRs merged together anyway.

I am not against small PR strategy, but in this case it is not really possible to avoid large PR in order for it to be testable.
Merging PRs that we know will fail is a dangerous and we should avoid it,

ca-scribner replied:
I agree that merging things that we know will fail **into main** is dangerous and we should not do that.  As discussed earlier today, that is not what is being discussed.  

What is being said is that, for things that can clearly be separated into independent concerns (eg: the groups @DnPlas mentioned) we should create separate PRs.  If these PRs do not result in a working final product on their own, **they should be opened against a feature branch**.  This way they can each be discussed on their own merits and merged when they're ready.

@jnsgruk you're right that we can separate these different concerns into different commits.  100% agree.  A reason why imo that is less effective than different PRs against a feature branch is because as soon as reviews start, those three separate commits can be updated, force pushed, etc.  Every time a reviewer comes back to the review, they need to interrogate the PR to understand what changed.  By keeping things in separate PRs, the effort on the reviewer is reduced and we can merge non-controversial things quickly.

i-chvets replied:
My point is simple, common approach (outlined in our best practices) is for PR to be testable (regardless of size). When we commit separate PRs that cannot be tested, I am not sure code review is valid. Reviewing code that is not working is wasting time, unless it is a draft. Because, later on, when everything is merged (all small PRs) new review needs to be done or when feature branch is merged into `main` someone will have to do it: *The sum of the parts does not equal the whole* 
We do not have formal QA that would verify final functionality for us. I doubt anyone goes and tests end-to-end functionality of final build of `main` branch.

beliaev-maksim replied:
@i-chvets 
going into the PR files I see the common base that we use (same as in other charms)
let's take concrete example:
1. renovate -> can easily go into another PR
2. readme, the same
3. tox.ini, the same
4. tests requirements , the same
5. dockerfile, the same
6. gitignore, the same
7. pyproject.toml, the same

these 7 items could be addressed before even writing the code.
and then you have a single PR with all the tests and python code, that is self contained and testable.

i-chvets replied:
1. Agreed.
2. Agreed (partially). However, to understand what I am reviewing in other small PRs I would need to look into README, thus switching PRs... deafting purpose of small PRs.
3. Disagree: Executing `tox` commands will fail without source code and tests. What am I reviewing then? It is nicely formatted `tox.ini` file, but does it work? I have no idea. Show me how I can test that `tox.ini` without source and tests, and I will switch disagree to agree. If this `tox.ini` is a boilerplate, let's automate it, so we don't have to review it. For now to test it, I need to bring in "source/test only" PR to verify that `tox.ini` is functional. Thus defeating the purpose of small PRs.
4. Disagree. requirements.* go with the code. If I need to bring some package that I need in my source code, I will have to commit requirements change along with the code, otherwise "source/test only" PR will fail and I won't be able to test it. I would have to bring PR with requirements into the fold, thus defeating the purpose of small PR.
5. Disagree (partially). Dockerfile is useless without the source code. I would have it in a separate PR for it at later stage (after source code is checked in), but at this point I would need to wait for all other PRs to be merged before reviewing Dockerfile.
6. Agree.
7. Agree.

ca-scribner commented on .github/workflows/on_push_pull_request_schedule.yaml:
Before having real code in the repo, it would be nice to have linting and testing CI set up.  It doesn't need to pass necessarily, but then we see red X's instead of green checkmarks that don't assert much.  

Adding the CI would be a great thing to break into a separate PR that gets merged first.  Or even just the lint+unit+integration tests combined with the `tox.ini` and whatever is needed.  Getting that boilerplate handled up front makes this review a bit easier

ca-scribner commented on server.py:
We need some tests on this function.  On the face it looks correct, but there's a lot going on that's hard to visually inspect.

Upstream's sync.py has [tests](https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/base/installs/multi-user/pipelines-profile-controller/test_sync.py) that might give some inspiration.  Not sure how applicable they are, but it at least is a start.

misohu replied:
added some

ca-scribner commented on server.py:
```diff
@@ -0,0 +1,128 @@
+        def sync(self, parent, children):
+            logger.info("Got new request")
+            context = {"namespace": parent.get("metadata", {}).get("name")}
+            pipeline_enabled = parent.get("metadata", {}).get("labels", {}).get(label)
+
+            if pipeline_enabled != "true":
```
Do you plan to use the pipelines label to identify which namespace is a kubeflow user, or will you add our own label?  Since it is an arg, I think you just mean label here rather than something about pipelines?

ca-scribner replied:
Ok then let's update the naming to reflect this.  The `pipeline_enabled` variable's name suggests we're only operating on namespaces that have pipelines enabled, but we handle whatever we've labelled as in scope

ca-scribner commented on server.py:
```diff
@@ -0,0 +1,128 @@
+                )
+                return {"status": {}, "children": []}
+
+            desired_secret_count = 0
```
`desired_secret_count = 0` says "my desired state is to have no secrets" - is this the intent?  I think this is going to tell metacontroller "if you have 0 secrets, dont ever change that" and then no items will ever be created.  

This probably should be set dynamically based on your manifests.  The upstream sync.py you're using as a guide could get away with a static value here because they knew what it would be, but in your case you do not

ca-scribner reviewed (commented):
This is going in a good direction, but hard to evaluate without tests.  Let's add some tests, get the CI to run them plus linting, etc, and then we can really evaluate this PR. 

For things like CI and the boilerplate (renovate.json, requirements*, tox, etc), break those out into a separate PR.  The CI will fail to start, but that's ok

ca-scribner commented:
I think adding structure to the repo might help here.  Having the build (dockerfile) and source all at the root level of the repo clutters things up.  Maybe adding a few subfolders for these things would help?

i-chvets commented on src/server.py:
```diff
@@ -0,0 +1,79 @@
+    for template_file in template_files:
+        template = Template(Path(template_file).read_text())
+        rendered_template = template.render(**context)
+        tmp_yaml = yaml.safe_load(rendered_template)
```
Should we add exception handling around file operations?

misohu replied:
Added one for template exception

misohu commented:
Ok fixed README and added exception for jinja template rendering. I also put the code for dispatcher into folder Based on @i-chvets I am keeping it in single PR ... @ca-scribner feel free to review again

ca-scribner commented on resource_dispatcher/tests/integration/test_server_controller.py:
```diff
@@ -0,0 +1,94 @@
+    "parent": {"metadata": {"name": "someName", "labels": {LABEL: "true"}}},
+    "children": {
+        "Secret.v1": [],
+        "ConfigMap.v1": [],
+        "Deployment.apps/v1": [],
+        "Service.v1": [],
```
copy and paste error?  
```suggestion
```

ca-scribner commented on resource_dispatcher/src/server.py:
```diff
@@ -0,0 +1,90 @@
+
+            desired_secret_count = 1
+            desired_resources = []
+            desired_status = {
+                "resources-ready": str(len(children["Secret.v1"]) == desired_secret_count)
+            }
```
This one is interesting.  I misunderstood this field until you used it here.  I thought it was a boolean "are my resources right?" sent to metacontroller, but it actually [sets the object's status field](https://metacontroller.github.io/metacontroller/api/compositecontroller.html#sync-hook-response) like you're doing here.  I like this addition.

When you get to charming this, keep this field in mind.  This is a lagging indicator (eg: this execution of the script returns {`status before I started working`, `resources I want to see`}.  So if we see things need updating, we're going to report that `status=bad`.  If metacontroller then calls us back a few seconds later once it built the resources we wanted, we'll respond `status=good` and all is well.  But the `CompositeController`'s [resync period](https://github.com/canonical/resource-dispatcher/pull/1/files#diff-900c2a543e4efb22c9953cba878be42fd0b65fb20f0c3cabcbc4568f6dc8e754R7) is what controls this.  10s like you have in the other PR right now is probably great (we can only be behind in status by 10s).  3600s like upstream uses would be bad.

ca-scribner replied:
What could help here is the optional third data item you can return, `resyncAfterSeconds`.  What you could do here is if `resources-ready==False`, add a `resyncAfterSeconds=10` or something similar to the response.  This'll tell metacontroller that you need a resync soon, and then our statuses will not be stale even if the CompositeController's default is a long sync process

ca-scribner replied:
We discussed offline.  This change was important because the script here has no control over the resync pattern that someone submits it with, so a user could (incorrecltly) deploy it with a long resync time.  

But this is now fixed by https://github.com/canonical/resource-dispatcher-image/pull/1/commits/eef10f913e701c93c26c9a3cbc2caf22488ad378 and https://github.com/canonical/resource-dispatcher-image/pull/1/commits/f23798d243b9d057789b3c2237dbe1fca3d4ca62

misohu commented:
I have made some small change ... I removed the jinja logic and replaced it with pure python dictionary insert as jinja is overkill for this use case .... we will always only need to fill the namespace section and the templating happens in charm controllers.

ca-scribner reviewed (approved):
lgtm, very good work!
